src/util/log.d

```diff
@@ -350,18 +354,71 @@ class FileLogger(alias Layout) : Logger
     }
 }
 
-class RollingFileLogger(alias Layout) : FileLogger!Layout
+mixin template CachedAppender()
+{
+    import std.process : thisThreadID;
+
+    private static struct ThreadState
+    {
+        Appender!(char[]) appender;
+
+        int depth;
+    }
+
+    private ThreadState[typeof(thisThreadID)] threadStates;
+
+    private ref ThreadState threadState() return
+    {
+        return this.threadStates.require(thisThreadID, ThreadState(.appender!(char[]), 0));
+    }
+
+    // Only use cached appender if we're not recursing (logging inside a toString).
+    private Appender!(char[]) getAppender()
+    {
+        synchronized (this)
+        {
+            with (threadState)
+            {
+                depth ++;
+                if (depth == 1)
+                {
+                    assert(appender.data.empty);
+                    return appender;
+                }
+            }
+            return .appender!(char[]);
```

Comment:
Why `.appender`? Outside `with`, no `.` should be required?

Reply:
removed